test/jdk/jdk/nio/zipfs/ZipFSTester.java

```diff
@@ -686,10 +686,22 @@ static void test8131067() throws Exception {
     private static FileSystem newZipFileSystem(Path path, Map<String, ?> env)
         throws Exception
     {
+        // For the path which is not encoded by UTF-8, we need to
+        // replace "+" by "%2b" manually before feeding into the decoder.
+        // Otherwise, the URLDecoder would replace "+" by " ", which may
+        // raise NoSuchFileException.
+        //
+        // eg. var path = "file:///jdk-18+9/basic.jar";
+        //     URLDecoder.decode(path, "utf8") -> file:///jdk-18 9/basic.jar
+        //
+        // Also, we should not use URLEncoder in case of the path has been
+        // encoded.
+        var encoded = path.toUri().toString().replace("+", "%2b");
+
         // Use URLDecoder (for test only) to remove the double escaped space
         // character
         return FileSystems.newFileSystem(
-            new URI("jar", URLDecoder.decode(path.toUri().toString(), "utf8"),
+            new URI("jar", URLDecoder.decode(encoded, "utf8"),
```

Comment:
instead of using local variable encoded, I would just add the call to .replace("+", "%2b") inline.

The comments look good, I would suggest to put them between line 702/703.

Reply:
Done. A new patch has been uploaded.